### PR TITLE
Logging ordering fixes and feedback

### LINKED ISSUE #4893: [AzureTools][Aspire] The sequence number of the log changes when returning "Resource" and then clicking "View" of the project or container under the "Logs" in the Dashboard again
**Update:** More is broken here than just the order of logs. See https://github.com/dotnet/aspire/issues/4893#issuecomment-2274700970 for details.

---

**Regression:** Not a regression, this issue also reproduces on VS 17.11.0 preview 3 with Aspire 8.0.0

**Clean machine:** Win11 x64 23h2 ENU
**VS Version:** VS 17.12.0 Preview 1.0 [35113.27.main]
**Aspire Version:** 8.1.0-preview.1.24362.2
Apply NuGet Feeds

**REPRO STEPS:**

1. Create a new .NET Aspire Starter Application with redis checked -> Run `dotnet run` in the AppHost.project folder.
2. Click "View" of the project or container under the "Logs" in the Dashboard.
3. Return to "Resource" -> Click "View" of the project or container under the "Logs" in the Dashboard again.
4. Check whether the sequence number of the log remains unchanged.

**Expect:**
The sequence number of the log remains unchanged.

**Actual:**
The sequence number of the log changes.
Actual1 gif

**More Info:**
This issue also **reproduces** for other projects.

### PR DESCRIPTION
**Pull request by JamesNK**
## Description

Following feedback from [the previous PR](https://github.com/dotnet/aspire/pull/5357) and discussion with Karol about what options we have for getting logs from DCP, it's clear that there isn't a robust, fast, simple way to stream correctly ordered console logs from the host. The only way to produce the correct order is to do a final sort on the client.

This PR:

* Removes flaky time based delay from the server. It isn't possible to perfectly order log lines before they are returned from the service. Don't attempt to delay sending while ordering.

* Shares code for how log entries are internally represented on the client (dashboard) and server (host). Parsing and sorting logic is shared to ensure the backlog and client results match in content and order.

* Improves how sorting is implemented. Log lines with dates are placed in sorted order, while those without dates are placed in the order they were received. This mades ordering somewhat complicated, but there are tests, and I think it's important for the dashboard to be resilient to receiving whatever data it gets given as logs. In saying that, I want to make sure DCP always includes log lines with a prefixed date.

* Changes ResourceLoggerService to use one lock. I didn't see the benefit of multiple. It makes locking easier to reason with and reduces the chance of accidential deadlock.

* Resource logger state backlog is cleared immediately when there are no longer any subscribers. Previously they were cleared via a notificated sent through a number of channels. A race condition could cause logs to be watched again before they were cleared. That would cause logs to be double returned: the backlog + all logs from the start again.

Fixes https://github.com/dotnet/aspire/issues/4893

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5398)

### CODE REVIEW

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 1]
```diff
@@ -494,13 +487,16 @@ private void StartLogStream<T>(T resource) where T : CustomResource
                     {
                         foreach (var (content, isError) in batch)
                         {
-                            if (!ResourceLoggerService.TryParseContentLineDate(content, out var dateTimeUtc))
+                            DateTime? timestamp = null;
+                            var resolvedContent = content;
+
+                            if (TimestampParser.TryParseConsoleTimestamp(resolvedContent, out var result))
```
Comment (eerhardt) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:493]: Why do we parse the timestamp out here, just to turn around and put them back together in the `ResourceLoggerService`? Why not just leave the timestamp in the content, and pass it all the way up to the dashboard. And only need to parse it out once in the UI?

File: src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs [hunk 2]
```diff
@@ -359,6 +331,11 @@ private event Action<InternalLogLine> OnNewLog
 
                 if (raiseSubscribersChanged)
                 {
+                    // Clear backlog immediately.
+                    // Avoids a race between message being subscription changed notification eventually clearing the
+                    // logs and someone else watching logs and getting the backlog + complete replay off all logs.
+                    ClearBacklog();
```
Comment (eerhardt) [inline src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs:337]: Isn't there still a race here? This happens outside of the lock. So we could release the lock above, and immediately a new subscriber could get added.

File: src/Shared/ConsoleLogs/LogEntries.cs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                if (missingTimestampIndex == null)
+                {
+                    missingTimestampIndex = rowIndex;
+                }
+
+                continue;
+            }
+
+            // Add log entry if it is later than current entry.
+            if (logEntry.Timestamp.Value >= current.Timestamp.Value)
```
Comment (karolz-ms) [inline src/Shared/ConsoleLogs/LogEntries.cs:89]: What if you receive a bunch of records with the same timestamp value? Presumably you want to insert them in the order received, but I do not think the current implementation will do that? A test for that would probably be handy.

Comment (JamesNK) [reply src/Shared/ConsoleLogs/LogEntries.cs:89]: Yes it inserts in order.

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 1 above)
Comment (JamesNK) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:493]: I did this so the content in the host LogEntry was consistent with the client LogEntry.

File: src/Shared/ConsoleLogs/LogEntries.cs (hunk 3 above)
Comment (JamesNK) [reply src/Shared/ConsoleLogs/LogEntries.cs:89]: Test added.

File: src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs (hunk 2 above)
Comment (JamesNK) [reply src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs:340]: Added lock around unsubscribe.

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 1 above)
Comment (eerhardt) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:493]: Why do we need them to be consistent? The host can just deal with the logs as lines of content.

it’s only the UI that wants them to be sorted by date time, so why would we force the host to parse the time out?

File: src/Shared/ConsoleLogs/TimestampParser.cs [hunk 4]
```diff
@@ -36,6 +34,7 @@ public static bool TryParseConsoleTimestamp(string text, [NotNullWhen(true)] out
                 }
             }
 
+            var timestamp = match.ValueSpan;
             result = new(content.ToString(), DateTimeOffset.Parse(timestamp.ToString(), CultureInfo.InvariantCulture));
```
Comment (eerhardt) [inline src/Shared/ConsoleLogs/TimestampParser.cs:38]:
```suggestion
            result = new(content.ToString(), DateTimeOffset.Parse(timestamp, CultureInfo.InvariantCulture));
```

No reason to create an intermediate string here.

**Review by karolz-ms — APPROVED** [review]

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs (hunk 1 above)
Comment (JamesNK) [reply src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:493]: They don't need to be consistent; I just think it's simpler to reason about everything when it's consistent and it allows more code to be shared between the host and dashboard.

In the host, either the `ILogger` logs need to add a timestamp to log line content or DCP logs need to remove a timestamp from log line content. Either way, it's slightly weird.

I chose to make the DCP logs remove the time stamp over the alternative for the reasons above. The downside there are some extra string allocations, the upside is storing the backlog uses less memory. Date a long `DateTime` is smaller than vs 15ish UTF16 chars in the log message.

File: src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs [hunk 5]
```diff
@@ -282,65 +280,44 @@ void Log(InternalLogLine log)
...
             {
-                bool raiseSubscribersChanged;
-                lock (this)
-                {
-                    raiseSubscribersChanged = _onNewLog is null; // is this the first subscriber?
+                Debug.Assert(Monitor.IsEntered(_lock));
 
-                    _onNewLog += value;
-                }
+                var raiseSubscribersChanged = _onNewLog is null; // is this the first subscriber?
```
Comment (drewnoakes) [inline src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs:318]: It's unclear why we only raise a subscriber-change notification for the first subscriber. A comment would be good.

**Review by drewnoakes — APPROVED** [review]

**Pull request merged**